Scripts for processing and visualizing data associated with reads from high-throughput sequencing
===============================================================================================

# The scripts

* plot_5prime_end_from_bedgraph.py
> 5'-end data in bedgraph form --> plot of 5'-ends for specified region (or expoerted relative data)

**IMPORTANTLY: There is a notebook demonstrating use of this script and illustrating many of the options [available here](https://github.com/fomightez/clausen_ribonucleotides).**

Verified compatible with both Python 2.7 and Python 3.6.

Written to run from command line or pasted/loaded inside a Jupyter notebook cell.  
The main ways to run the script are demonstrated in [a series of notebooks illustrating using PatMatch with Python](https://github.com/fomightez/patmatch-binder).

USAGE

Example calls to run script from command line:
```
python plot_5prime_end_from_bedgraph.py 5013-5359 forward.bedgraph reverse.bedgraph

python plot_5prime_end_from_bedgraph.py 5013-5359 forward.bedgraph reverse.bedgraph --sample_id WT3b

python plot_5prime_end_from_bedgraph.py 5013-5359 forward.bedgraph reverse.bedgraph --sample_id WT3b --reverse_orientation
```

(Alternatively, upload the script to a Jupyter environment and use `%run plot_5prime_end_from_bedgraph.py 5013-5359 forward.bedgraph reverse.bedgraph` in a Python-backed notebook to run the example.)

* plot_coverage.py (NEED TO PUT IN THIS REPO STILL)
> high-throughput sequencing data --> plot of coverage for specified region (or expoerted relative data)

Takes data previously generated by the use of [Samtools](http://www.htslib.org/doc/samtools.html) with sorted and indexed BAM files for high-throughput sequencing data (intended for Illumina-generated data) and makes a plot of the coverage for the specified region. It can also be used to pass data from that source to other computational tools. 

* plot_coverage_and_starts.pyy (NEED TO PUT IN THIS REPO STILL)
> high-throughput sequencing data --> plot of coverage and 5'-ends for specified region (or expoerted relative data)

Takes data previously generated by the use of [Samtools](http://www.htslib.org/doc/samtools.html) with sorted and indexed BAM files for high-throughput sequencing data (intended for Illumina-generated data) and makes a plot of the coverage and 5'-ends for the specified region. It can also be used to pass data from that source to other computational tools. 

**Note for plot_coverage.py and plot_coverage_and_starts.py  I intend to put here-->** These need data previously generated by the use of [Samtools](http://www.htslib.org/doc/samtools.html) with sorted and indexed BAM files for sequencing data.  
**Another Note for the  plot_coverage.py and plot_coverage_and_starts.py  I intend to put here first -->** Despite the names, they both have `--export_data` options that can be called to easily pass the data they mine from the input data, but **WHEN THIS OPTION IS USED NO PLOT IS MADE**. It is meant to be used in conjunction with plots probably already made and with other scripts, see the section `Related items created by myself` below. It was a deliberate choice to add directly to the script the ability to export the data rather than duplicating the data processing steps in a separate script with the idea it would lessen the need for propagating changes to multiple places to keep such related scripts in sync.

Related items created by myself
-------------------------------

I have made a script that utilizes the `--export_data` option to run `plot_coverage_and_starts.py` to automate an effort:
* [`shepherds_read_starts_at_start_of_origins_thru_to_plotting.py`](https://github.com/fomightez/mini-pipelines)

Specifically, the script [`shepherds_read_starts_at_start_of_origins_thru_to_plotting.py`](https://github.com/fomightez/mini-pipelines) utilizes the `--export_data` option to run `plot_coverage_and_starts.py` in an automated manner and plot the starts in a subregion of the data using the script `plot_stacked_bar_plots_with_fit.py`. The associated script `plot_stacked_bar_plots_with_fit.py` is found [here](https://github.com/fomightez/general_scripted_plotting), as it is general and not specific to read data.


Related items by others
-----------------------

Not for high-throughput data, but for a different type of modern sequencing data:

[NanoPack - NanoPlot and NanoStats and other Python scripts and modules for visualizing and processing Oxford Nanopore sequencing data](https://github.com/wdecoster/nanopack)
